std/haxe/Serializer.hx

```diff
@@ -433,11 +427,9 @@ class Serializer {
 					serializeFields(v);
 				}
 			case TEnum(e):
-				if (useCache) {
-					if (serializeRef(v))
-						return;
-					cache.pop();
```

Comment:
This however is relevant in order to preserve the correct order. In case of nested enums, we deal with the inner ones first when deserializing, which inverts the order. Here's a failing example:

```haxe
enum E {
	B;
	C(e1:E, e2:E);
}

function main() {
	Serializer.USE_CACHE = true;
	var e = C(B, B);
	var s = Serializer.run(e);
	trace(s); // wy1:Ey1:C:2wR0y1:B:0r1
	var e2 = Unserializer.run(s);
	trace(e2);
}
```

Note the `r1` reference, which fails because by the time we're seeing it, only `B` has been committed and would be `r0` instead.